[bytecode verifier] Improve performance of stacks

tnowacki opened the pull request:
## Description 

- Reduce memory allocation for abstract stacks in the verifier
- Certain operations can introduce multiple copies of the same type on the stack. This run-length style encoding will reduce allocation in those cases 

## Test Plan 

- Non-functional change, used existing tests 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

dariorussi reviewed (commented):
This looks pretty good to me!
what do we want to do here?
I'd be happy to write a bunch of tests on the specific stack for you if you'd like me to.
Also where and when do we want to push it?
Should we start versioning the VM? do we want to get this in quickly?
let me know plwase

tnowacki commented on external-crates/move/move-bytecode-verifier/src/abstract_stack.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /// Pops a single value off the stack, erroring if empty
+    pub fn pop(&mut self) -> Result<T, AbsStackError> {
```
I think that would work, yeah.

tnowacki commented on external-crates/move/move-bytecode-verifier/src/abstract_stack.rs:
```diff
@@ -0,0 +1,106 @@
+
+    /// Push n copies of an item on the stack
+    pub fn push_n(&mut self, item: T, n: u64) {
+        if n == 0 {
```
It forces a guard then on each usage and just made things look a bit more gross for cases like `VecUnpack` when the vector was empty. 
I can switch it if you feel strongly though.

dariorussi replied:
I see and good point, did not notice the function does not return an error.
I do not feel strong at all, and the "seriously?" was referred to a conceptual users doing that operation rather than my feelings on the subject which are pretty tame...

tnowacki commented on external-crates/move/move-bytecode-verifier/src/abstract_stack.rs:
```diff
@@ -0,0 +1,106 @@
+        match self.values.last_mut() {
+            Some((count, last_item)) if &item == last_item => {
+                debug_assert!(*count > 0);
+                *count += n
```
Accounting for the true overflow here seems like some worthwhile hardening. 

I'm not sure about the logical case though, especially since we don't expose a size/len function.

tnowacki commented:
> This looks pretty good to me! what do we want to do here? I'd be happy to write a bunch of tests on the specific stack for you if you'd like me to. 
I came here looking for suggestions too :) 

I'm thinking at least we should write some stand alone unit tests for the stack itself. Past that, I'm a bit less worried about tests for the specific passes. Or maybe this is what you had in mind?

> Also where and when do we want to push it? Should we start versioning the VM? do we want to get this in quickly? let me know plwase

Thoughts @amnn?

amnn commented:
I keep getting sidetracked by planning :( I think the question of how quickly this should go in is the main decider here -- if it can wait a week, it can be versioned, if not then we should consider other options.

dariorussi commented:
> I'm thinking at least we should write some stand alone unit tests for the stack itself. Past that, I'm a bit less worried about tests for the specific passes. Or maybe this is what you had in mind?

yes I had in mind tests for this specific piece of code. I'll go ahead and made some, sounds good?

tnowacki commented:
@dariorussi, this PR has changed a good deal:
- pop_n API has changed into EQ and non-EQ  variants
- I added a len() API in order to fix some weird cases where modifications were made even on errors
- push now gives a result to deal with u64 overflows. It's a bit annoying, but makes sense I think given how easy it is to get to u64::max with the push_n API 

I added some extensive unit tests, which should make this a bit easier to go in with confidence?

dariorussi commented on external-crates/move/move-abstract-stack/src/unit_tests/mod.rs:
```diff
@@ -0,0 +1,265 @@
+    s.push(3).unwrap();
+    s.push(3).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 6);
```
wonder if there was an easy way in this test to get into the internal of the stack and check that "compaction" had happened....

tnowacki replied:
It's a real good question. I'll poke at it

dariorussi reviewed (approved):
this is great! love it!
Quick comment, if at some point we wanted to add a size to the stack what would be your take/reaction? 
Basically not just an overflow for `u64::MAX` but for a given value....

tnowacki commented on external-crates/move/move-abstract-stack/src/lib.rs:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    /// Pop any n items off the stack. Unlike `pop_n`, items do not have to be equal
+    pub fn pop_any_n(&mut self, n: NonZeroU64) -> Result<(), AbsStackError> {
```
I agree as it would be really strange to leave it partially updated.
I changed it last commit to check for the length first and to error without mutations if there are not enough values

tnowacki commented on external-crates/move/move-abstract-stack/src/lib.rs:
```diff
@@ -0,0 +1,163 @@
+    pub fn pop_eq_n(&mut self, n: NonZeroU64) -> Result<T, AbsStackError> {
+        let n: u64 = n.get();
+        if self.is_empty() {
+            return Err(AbsStackError::EmptyStack);
```
So would `ElementNotEqual` and `NotEnoughValues` and `EmptyStack` all be `Underflow`?

tzakian commented on external-crates/move/move-abstract-stack/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+
+#[derive(Default, Debug)]
+/// An abstract value that compresses runs of the same value to reduce space usage
+pub struct AbsStack<T> {
```
Thoughts on calling this `AbstractStack`? Or maybe `RLEStack`? `AbsStack` feels like a needless abbreviation to me/a bit jarring when reading the code.

tzakian commented on external-crates/move/move-bytecode-verifier/src/type_safety.rs:
```diff
@@ -809,17 +823,21 @@ fn verify_instr(
+                    .stack
+                    .pop_eq_n(num_to_pop)
+                    .map(|t| element_type != &t)
+                    .unwrap_or(false);
```
I think this is wrong? I think this should be an `unwrap_or(true)`.

If there are only `k < n` types in `AbsStack` then `pop_eq_n` will return an `Err(AbsStackError::ElementNotEqual)`

This will then get unwrapped to `false` which will then not trigger the below if statement.

tnowacki replied:
How do we not have a failing test for this...

tnowacki replied:
How do we not have a test for this... good catch! great catch!

tnowacki replied:
Added a test, good catch! Great catch!

tzakian commented on external-crates/move/move-bytecode-verifier/src/type_safety.rs:
```diff
@@ -863,20 +881,18 @@ fn verify_instr(
-            for _ in 0..*num {
-                verifier.push(meter, declared_element_type.clone())?;
-            }
+            verifier.push_n(meter, declared_element_type.clone(), *num)?;
```
This changes metering behavior from `ty_push_cost * n` to just `ty_push_cost`. re we ok with this change? I'd be more comfortable with charging ` * n` if possible...

tnowacki replied:
As in keeping the behavior from before the PR

tzakian commented on sui-execution/latest/sui-verifier/src/id_leak_verifier.rs:
```diff
@@ -409,14 +427,14 @@ fn execute_inner(
         // These bytecodes produce references, and hence cannot be ID.
         Bytecode::MutBorrowLoc(_)
-        | Bytecode::ImmBorrowLoc(_) => verifier.stack.push(AbstractValue::Other),
+        | Bytecode::ImmBorrowLoc(_) => verifier.stack_push(AbstractValue::Other).unwrap(),
```
Any reason this should be an unwrap as opposed to a `?`

tnowacki replied:
I think I missed just switching this to ? (which wasn't possible in an earlier version of this PR)

tzakian reviewed (changes requested):
Overall looks good to me! Left a couple nits, but requesting changes due to the possible issues in `type_safety` that I commented on.

tnowacki commented on external-crates/move/move-abstract-stack/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+                }
+                Ordering::Greater => {
+                    *count -= rem;
+                    rem = 0;
```
Always torn on these things. (editing the control flow vs a more declarative "take all of the rem")
But I'll switch to break :)

tnowacki commented on external-crates/move/move-abstract-stack/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+
+    /// Returns the logical length of the stack as if there was no space
+    /// compression
+    pub fn len(&self) -> u64 {
```
Sure, but `len` is more consistent with rust naming conventions (which uses len for all data structures)

tzakian reviewed (approved):
Looks great. Thanks for making those changes!
